runtime/floats.c

```diff
@@ -170,6 +163,84 @@ CAMLexport void caml_Store_double_array_field(value val, mlsize_t i, double dbl)
 }
 #endif /* ! FLAT_FLOAT_ARRAY */
 
+#ifdef HAS_VSNPRINTF_L
+static value caml_alloc_sprintf_l(locale_t loc, const char * format, ...)
+{
+  va_list args;
+  char buf[128];
+  int n;
+  value res;
+
+#ifndef _WIN32
+  /* BSD-compatible implementation */
+  va_start(args, format);
+  /* "vsnprintf_l(dest, sz, loc, format, args)" writes at most "sz" characters
+     into "dest", including the terminating '\0'.
+     It returns the number of characters of the formatted string,
+     excluding the terminating '\0'. */
+  n = vsnprintf_l(buf, sizeof(buf), loc, format, args);
+  va_end(args);
+  if (n < sizeof(buf)) {
+    /* All output characters were written to buf, including the
+       terminating '\0'.  Allocate a Caml string with length "n" as
+       computed by vsnprintf_l, and copy the output of vsnprintf_l into it. */
+    res = caml_alloc_initialized_string(n, buf);
+  } else {
+    /* PR#7568: if the format is in the Caml heap, the following
+       caml_alloc_string could move or free the format.  To prevent
+       this, take a copy of the format outside the Caml heap. */
+    char * saved_format = caml_stat_strdup(format);
+    /* Allocate a Caml string with length "n" as computed by vsnprintf_l. */
+    res = caml_alloc_string(n);
+    /* Re-do the formatting, outputting directly in the Caml string.
+       Note that caml_alloc_string left room for a '\0' at position n,
+       so the size passed to vsnprintf_l is n+1. */
+    va_start(args, format);
+    vsnprintf_l((char *)String_val(res), n + 1, loc, saved_format, args);
+    va_end(args);
+    caml_stat_free(saved_format);
+  }
+  return res;
+#else
+  /* Implementation specific to the Microsoft UCRT/CRT library */
+  va_start(args, format);
+  /* "_vsnprintf_l(dest, sz, format, loc, args)" writes at most "sz" characters
+     into "dest".  Let "len" be the number of characters of the formatted
+     string.
+     If "len" < "sz", a null terminator was appended, and "len" is returned.
+     If "len" == "sz", no null termination, and "len" is returned.
+     If "len" > "sz", a negative value is returned. */
+  n = _vsnprintf_l(buf, sizeof(buf), format, loc, args);
+  va_end(args);
+  if (n >= 0 && n <= sizeof(buf)) {
+    /* All output characters were written to buf.
+       "n" is the actual length of the output.
+       Allocate a Caml string of length "n" and copy the characters into it. */
+    res = caml_alloc_string(n);
```

Comment:
Why not use `caml_alloc_initialized_string` here too?

Reply:
Good catch - it appears to have been missed in #1274 when `caml_alloc_initialized_string` was introduced!